benchopt/stopping_criterion.py

```diff
@@ -0,0 +1,192 @@
+
+    This base class will check for the timeout and the max_run.
+    It should be sub-classed to check for the convergence of the algorithm.
+
+    This class also handle the detection of diverging solvers and print the
+    progress if given a ``prgress_str``.
+    """
+
+    def __init__(self, *args, **kwargs):
+        pass
+
+    @classmethod
+    def _get_instance(cls, *args, max_runs=1, timeout=None, progress_str=None,
+                      **kwargs):
+        # Instanciate the subclass
+        stopping_criterion = cls(*args, **kwargs)
+
+        # Store critical parameters for hashing and reconstruction
+        stopping_criterion.timeout = timeout
+        stopping_criterion.max_runs = max_runs
+        stopping_criterion.args = args
+        stopping_criterion.kwargs = kwargs
+
+        # Store running arguments
+        if timeout is not None:
+            stopping_criterion._deadline = time.time() + timeout
+        else:
+            stopping_criterion._deadline = None
+        stopping_criterion._prev_objective_value = 1e100
+        stopping_criterion._progress_str = progress_str
+        return stopping_criterion
+
+    def should_stop_solver(self, cost_curve):
+        """Base call to check if we should stop running a solver.
+
+        This base call check for the timeout and the max number of runs.
+        It also notify the runner if the curve is too flat, to increase
+        the number of points between 2 evaluations of the objective.
+
+        Parameters
+        ----------
+        cost_curve : list of dict
+            List of dict containing the values associated to the objective at
+            each evaluated points.
+
+        Returns
+        -------
+        stop : bool - wether or not we should stop the algorithm.
+        status : str - reason why the algorithm was stopped if stop is True.
+        is_flat : bool - indicate if 2 successive evaluation yielded the same
+            objective value. Useful for moving faster on tolerance.
+        """
+        # Modify the criterion state:
+        # - compute the number of run with the curve. We need to remove 1 as
+        #   it contains the initial evaluation.
+        # - compute the delta_objective for debugging and stalled progress.
+        n_eval = len(cost_curve) - 1
+        objective_value = cost_curve[-1]['objective_value']
+        delta_objective = self._prev_objective_value - objective_value
+        self._prev_objective_value = objective_value
+
+        # default for value for is_flat
+        is_flat = False
+
+        # check the different conditions:
+        #     timeout / max_runs / diverging / stopping_criterion
+        if self._deadline is not None and time.time() > self._deadline:
+            stop = True
+            status = 'timeout'
+
+        elif n_eval == self.max_runs:
+            stop = True
+            status = 'max_runs'
+
+        elif delta_objective < -1e10:
+            stop = True
+            status = 'diverged'
```

Comment:
This says that the objective should be decreasing, otherwise the solver is not converging and there is not point to keep it going (if I get it right).
But what about cases like with logreg_l2's  `copt[accelerated=True,llinesearch=False,solver=pgd]`  ([see on benchopt.results](https://benchopt.github.io/results/benchmark_logreg_l2_benchopt_run_2021-03-18_02h11m46.html) with the `simulated[200, 500]` dataset, the purple curve) where the objective increases at one point but the solver still converges at the end (also happens with [cvxpy in logreg_l1](https://benchopt.github.io/results/benchmark_logreg_l1_benchopt_run_2020-12-20_18h19m52.html) in the first iterations)

Reply:
Here, we are checking if the objective exploded (increased of 10^10). So it should only kick in in really bad situation I think. But let me know if you see this kick in some cases?

Reply:
Yes, sorry, I read it with ` e-10`  (not the same quantity at all ^^' ) so it's all good

Reply:
To be completely fair, I did the same multiple time :grin: